core/src/typecheck/mod.rs

```diff
@@ -1589,6 +1590,13 @@ fn check<L: Linearizer>(
             );
             unify(state, &ctxt, ty, ty_import).map_err(|err| err.into_typecheck_err(state, rt.pos))
         }
+        Term::Types(typ) => {
+            if let Some(flat) = typ.find_flat() {
+                Err(TypecheckError::FlatTypeInTermPosition { flat, pos: *pos })
+            } else {
+                Ok(())
```

Comment:
I was going to say that you might need to `walk` the type at least there, but this may not be true since we know it doesn't contain any flat type, so `walk` is probably useless.